File: chapter_5/Moody/moody.py
from abc import ABCMeta, abstractproperty


class Moody(metaclass=ABCMeta):
    @abstractproperty
    def mood():
        raise Exception('mood not found!')

    def query_mood(self):
        print(f"I'm feel {self.mood} today!!!")


class Happy(Moody):
    @property
    def mood(self):
        return 'happy'

    def laugh(self):
        print('jkjdhhhhqiwhdqdui')


class Sad(Moody):
    @property
    def mood(self):
        return 'sad'

    def cry(self):
        print('wahhhhhhh boooo hoooo')


class Psychiatrist:
    def examine(self, patient: Moody):
        print('DOCTOR: How d u feel ?')
        print('PATIENT:')
        patient.query_mood()

    def observe(self, patient: Moody):
        if isinstance(patient, Happy):
            patient.laugh()
        else:
            patient.cry()

        print(f'This patient is {patient.mood}')

    def comment(self):
        print('U are crazy!')

File: chapter_4/Moody/Moody.java
package chapter_4.Moody;

public abstract class Moody {
    protected abstract String getMood();

    public void queryMood() {
        System.out.println("I feel " + getMood() + " today!");
    }
}

public class Sad() {
    protected String getMood() {
        return "sad";
    }

    public void cry() {
        System.out.println("boo hoo");
    }
}

public class Happy() {
    protected String getMood() {
        return "happy";
    }

    public void laugh() {
        System.out.println("hahahahaha");
    }
}